File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def check_sub(p, q):
            if not p and not q:
                return True

            if p and not q or q and not p:
                return False

            if p.val != q.val:
                return False

            return check_sub(p.left, q.left) and check_sub(p.right, q.right)

        def dfs(root):
            if not root:
                return False

            if check_sub(root, subRoot):
                return True
            
            return dfs(root.left) or dfs(root.right)


        return dfs(root)

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-4.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func goodNodes(root *TreeNode) int {
    if root == nil {
        return 0
    }
    
    type pair struct {
        node *TreeNode
        maxVal int
    }

    queue := []pair{{root, math.MinInt32}}
    res := 0

    for len(queue) > 0 {
        node, maxVal := queue[0].node, queue[0].maxVal
        queue = queue[1:]

        if node.Val >= maxVal {
            res += 1
        }

		newMax := maxVal
        if node.Val > newMax {
			newMax = node.Val
		}

        if node.Left != nil { 
            queue = append(queue, pair{node.Left, newMax})
        }

        if node.Right != nil { 
            queue = append(queue, pair{node.Right, newMax})
        }
    }

    return res
}
